Source/Particles/ParticleBoundaryBuffer.H

```diff
@@ -0,0 +1,102 @@
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef PARTICLEBOUNDARYBUFFER_H_
+#define PARTICLEBOUNDARYBUFFER_H_
+
+#include "Particles/ParticleBuffer.H"
+#include "Particles/MultiParticleContainer_fwd.H"
+
+#include <vector>
+
+/**
+ *  This stores particles that have left / been absorbed by domain and embedded boundaries.
+ */
+class ParticleBoundaryBuffer
+{
+public:
+    ParticleBoundaryBuffer ();
+
+    int numSpecies() const { return getSpeciesNames().size(); }
+
+    const std::vector<std::string>& getSpeciesNames() const {
+        static bool initialized = false;
+        if (!initialized)
+        {
+            amrex::ParmParse pp_particles("particles");
+            pp_particles.queryarr("species_names", m_species_names);
+            initialized = true;
+        }
+        return m_species_names;
+    }
+
+    void gatherParticles (MultiParticleContainer& mypc,
+                          const amrex::Vector<const amrex::MultiFab*>& distance_to_eb);
+
+    void clearParticles ();
+
+    void printNumParticles () const;
+
+    constexpr int numBoundaries () const {
+        return AMREX_SPACEDIM*2
+#ifdef AMREX_USE_EB
+            + 1
+#endif
+            ;
+    }
+
+private:
+    // over boundary, then number of species
+    std::vector<std::vector<ParticleBuffer::BufferType<amrex::PinnedArenaAllocator> > > m_particle_containers;
+
+    // over boundary, then number of species
+    std::vector<std::vector<int> > m_do_boundary_buffer;
+
+    mutable std::vector<std::string> m_species_names;
+};
+
+struct IsOutsideDomainBoundary {
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_plo;
+    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> m_phi;
+    int m_idim;
+    int m_iside;
+
+    template <typename SrcData>
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    int operator() (const SrcData& src,
+                    int ip, const amrex::RandomEngine& /*engine*/) const noexcept
+    {
+        const auto& p = src.getSuperParticle(ip);
+        if (m_iside == 0) {
+            if (p.pos(m_idim) < m_plo[m_idim]) { return 1; }
+        } else {
+            if (p.pos(m_idim) >= m_phi[m_idim]) { return 1; }
+        }
+        return 0;
+    }
+};
+
+struct CopyAndTimestamp {
+    int m_index;
+    int m_step;
+
+    template <typename DstData, typename SrcData>
+    AMREX_GPU_HOST_DEVICE
+    void operator() (const DstData& dst, const SrcData& src,
+                     int src_i, int dst_i) const noexcept
+    {
+        dst.m_aos[dst_i] = src.m_aos[src_i];
+        for (int j = 0; j < SrcData::NAR; ++j)
+            dst.m_rdata[j][dst_i] = src.m_rdata[j][src_i];
+        for (int j = 0; j < src.m_num_runtime_real; ++j)
+            dst.m_runtime_rdata[j][dst_i] = src.m_runtime_rdata[j][src_i];
+        for (int j = 0; j < src.m_num_runtime_int; ++j)
+            dst.m_runtime_idata[j][dst_i] = src.m_runtime_idata[j][src_i];
+        dst.m_runtime_idata[m_index][dst_i] = m_step;
+    }
+};
```

Comment:
No need for this to be in the header, since it's only used in the one source file. That'll just be more to compile for every file that includes this